app/src/main/java/com/tatar/currencify/ui/feature/ConvertedCurrenciesAdapter.kt

```diff
@@ -0,0 +1,137 @@
+) : RecyclerView.Adapter<ConvertedCurrenciesAdapter.ViewHolder>() {
+
+    var convertedCurrencies: List<ConvertedCurrency> = emptyList()
+        set(value) {
+            DiffUtil.calculateDiff(
+                ConvertedCurrenciesDiffUtil(convertedCurrencies, value)
+            ).dispatchUpdatesTo(this)
+
+            field = value
+        }
+
+    private lateinit var itemClickListener: ItemClickListener
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val layoutInflater = LayoutInflater.from(parent.context)
+        val binding = LayoutConvertedCurrencyRowBinding.inflate(layoutInflater, parent, false)
+        val viewHolder = ViewHolder(binding)
+
+        setItemClickListener(viewHolder)
+
+        return viewHolder
+    }
+
+    private fun setItemClickListener(viewHolder: ConvertedCurrenciesAdapter.ViewHolder) {
+        viewHolder.itemView.setOnClickListener {
+            val position = viewHolder.adapterPosition
+            if (position == -1) return@setOnClickListener
+
+            itemClickListener.onItemClick(convertedCurrencies[position].currencyCode)
+        }
+    }
+
+    override fun getItemCount(): Int = convertedCurrencies.size
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        holder.bind(convertedCurrencies[position])
+    }
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int, payloads: MutableList<Any>) {
+        if (payloads.isNullOrEmpty()) {
+            onBindViewHolder(holder, position)
+        } else {
+            with(payloads.first()) {
+                if (this is Bundle) {
+                    val currencyCode = getString(KEY_CURRENCY_CODE)
+                    val amount = getString(KEY_AMOUNT)
+
+                    with(holder.binding) {
+                        if (currencyCode != null) currencyAbbTv.text = currencyCode
+                        if (amount != null) currencyAmountEt.setText(amount)
+                    }
+                }
+            }
+        }
+    }
+
+    fun setItemClickListener(itemClickListener: ItemClickListener) {
+        this.itemClickListener = itemClickListener
+    }
+
+    inner class ViewHolder(
+        val binding: LayoutConvertedCurrencyRowBinding
+    ) : RecyclerView.ViewHolder(binding.root) {
+        fun bind(convertedCurrency: ConvertedCurrency) {
+            with(binding) {
+                currencyIv.setImageResource(convertedCurrency.currencyCode.toCurrencyFlag())
+                currencyAbbTv.text = convertedCurrency.currencyCode
+                currencyNameTv.text = convertedCurrency.currencyCode.toCurrencyString()
+                currencyAmountEt.setText(String.format("%.3f", convertedCurrency.amount))
+            }
+        }
+    }
+}
+
+interface ItemClickListener {
+    fun onItemClick(baseCurrency: String)
+}
+
+class ConvertedCurrenciesDiffUtil(
+    private val oldList: List<ConvertedCurrency>,
```

Comment:
Aren't `ConvertedCurrency` list always have unique elements? Maybe you can use `Set` and you comparisons would be better. E.g. in `areItemsTheSame`. Or for you the order is important somehow?

Reply:
This is an Android specific thing. `ConvertedCurrenciesDiffUtil` is used to optimise the performance of the `Recyclerview`. The point is comparing the old and new list items and change only the necessary fields in the `Recyclerview`.